refactor(chunkserver): Split OutputBuffer inner buffer

dmga44 opened the pull request:
The main goal of this change is to prepare the `OutputBuffer` to emplace
bigger pieces of data at once during pread calls. This could not be
achieved previously because the buffer should have a sequence of
header, CRC, data block, header, CRC and so on.

The class `OutputBuffer::Buffer` is created for this purpose. This is a
generic class to implement the three types of buffers an `OutputBuffer`
has. Taking this into consideration, the implementation of the
`OutputBuffer` could be made simpler and more readable.

Side changes:
- increase readability of putting data into the `OutputBuffer`.
- save the allocation of a block of zeros when asking for block beyond current number of blocks in chunk.
- modernize the `WriteStatus` enum.
- change `BuffersPool` to adapt to the new format of the `OutputBuffer`. Keep track of the number of block in the buffers instead of the number of buffers, and log it from time to time.
- add documentation for the `OutputBuffer` class.
- improve the unit tests suite of the `OutputBuffer` class.

The idea is to squash all these commits.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
refactor: Split OutputBuffer inner buffer

The main goal of this change is to prepare the OutputBuffer to emplace
bigger pieces of data at once during pread calls. This could not be
achieved previously because the buffer should have a sequence of
header, CRC, data block, header, crc and so on.

This commit also adds a little bit more of readability when putting
data into the OutputBuffer.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat: Save allocation of block of zeros

When asking for a block beyond the current number, the previous
implementation was allocating the full block of zeros (while creating
the vector) and then copying it. The proposed change saves the
allocating by directly using memset in the memory.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
fix: Fix outputBuffer unittest

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
refactor: Create OutputBuffer::Buffer class

This commit mainly creates the class OutputBuffer::Buffer. This is a
generic class to implement the three types of buffers an OutputBuffer
has. Taking this into consideration, the implementation of the
OutputBuffer could be made simpler and more readable.

As a side change, the WriteStatus enum modernized to use the enum
class.

Signed-off-by: Dave <dave@leil.io>

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -1582,11 +1622,11 @@ void ChunkserverEntry::writeToSocket() {
 					"New bytes in pack->outputBuffer after sending some data");
 			stats_bytesout += (bytesInBufferBefore - bytesInBufferAfter);
-			if (ret == OutputBuffer::WRITE_ERROR) {
+			if (ret == OutputBuffer::WriteStatus::Error) {
```
All sites using these write statuses were updated.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -1582,11 +1622,11 @@ void ChunkserverEntry::writeToSocket() {
 				state = State::Close;
 				return;
-			} else if (ret == OutputBuffer::WRITE_AGAIN) {
+			} else if (ret == OutputBuffer::WriteStatus::Again) {
```
Same as before.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -137,19 +137,18 @@ MessageSerializer *MessageSerializer::getSerializer(PacketHeader::Type type) {
+	outPacket->outputBuffer = getReadOutputBufferPool().get(packetPrefix.size(), numBlocks);
 
-	if (outPacket->outputBuffer->copyIntoBuffer(packetPrefix) !=
+	if (outPacket->outputBuffer->copyIntoBuffer(OutputBuffer::BufferType::Header, packetPrefix) !=
```
It is.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -361,49 +401,51 @@ void ChunkserverEntry::readContinue() {
+			std::vector<uint8_t> readDataPrefix;
+			messageSerializer->serializePrefixOfCstoclReadData(readDataPrefix, chunkId, offset,
+			                                                   thisPartSize);
+			auto packet = createDetachedPacketWithOutputBuffer(readDataPrefix, numBlocks);
```
It is the correct way of calculating it.

dmga44 commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -744,24 +748,26 @@ int hddRead(uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
+				    &crcBuffPointer,
+				    mycrc32(0, tmp.rawData(OutputBuffer::BufferType::Block) + offsetWithinBlock,
+				            size));
+				outputBuffer->copyIntoBuffer(OutputBuffer::BufferType::CRC, crcBuff, kCrcSize);
```
It is.

dmga44 commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -744,24 +748,26 @@ int hddRead(uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
+				outputBuffer->copyIntoBuffer(OutputBuffer::BufferType::CRC, crcBuff, kCrcSize);
 				outputBuffer->copyIntoBuffer(
-				    tmp.data() + kCrcSize + offsetWithinBlock, size);
+				    OutputBuffer::BufferType::Block,
+				    tmp.rawData(OutputBuffer::BufferType::Block) + offsetWithinBlock, size);
```
It is.

dmga44 commented on src/chunkserver/bgjobs.cc:
```diff
@@ -277,6 +277,8 @@ void* job_worker(void *th_arg) {
 						rdargs->offset, rdargs->size, rdargs->maxBlocksToBeReadBehind,
 						rdargs->blocksToBeReadAhead, rdargs->outputBuffer);
 
+				rdargs->outputBuffer->isReadCompleted = true;
```
It is the intended behavior, plus, this change belongs to another PR.

dmga44 commented on src/chunkserver/buffers_pool.h:
```diff
@@ -47,43 +47,49 @@ class BuffersPool {
-
-		if (buffers_.size() < kMaxSize) {
+		std::unique_lock lock(mutex_);
+		auto &buffers = buffersMap_[buffer->type()];
```
The `type()` means that the `headerSize` and `numBlocks` are correct.

dmga44 commented on src/chunkserver/chunkserver_entry.cc:
```diff
@@ -289,31 +288,78 @@ void ChunkserverEntry::delayedCloseCallback(uint8_t status, void *entry) {
 		eptr->isChunkOpen = 0;
 	}
-	eptr->state = State::Closed;
+	if (eptr->pendingReadJobIds.empty() && eptr->toDiscardReadJobIds.empty() && eptr->writeJobWriteId == 0) {
+		eptr->state = State::Closed;
```
Belongs to another PR.

dmga44 commented on src/chunkserver/output_buffer.h:
```diff
@@ -32,46 +32,123 @@
-
-		return capacity + disk::kIoBlockSize - remainder;
-	}
+	std::atomic_bool isReadCompleted{false};
```
Yes.

dmga44 commented on src/chunkserver/output_buffer_unittest.cc:
```diff
@@ -42,17 +43,23 @@ TEST(OutputBufferTests, outputBuffersTest) {
 	uint8_t buf[WRITE_SIZE];
 	memset(buf, VALUE, WRITE_SIZE);
-	ASSERT_EQ(outputBuffer.copyIntoBuffer(buf, WRITE_SIZE), WRITE_SIZE);
+	ASSERT_EQ(outputBuffer.copyIntoBuffer(OutputBuffer::BufferType::Header,
+	                                      std::vector<uint8_t>(testHeaderSize, VALUE)),
+	          testHeaderSize);
+	ASSERT_EQ(outputBuffer.copyIntoBuffer(OutputBuffer::BufferType::CRC, buf, kCrcSize), kCrcSize);
+	ASSERT_EQ(outputBuffer.copyIntoBuffer(OutputBuffer::BufferType::Block, buf, WRITE_SIZE_DATA),
+	          WRITE_SIZE_DATA);
```
Indeed.

dmga44 committed:
docs: Document the OutputBuffer class

This commit introduces documentation for the OutputBuffer class and the
related functions.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
test: Improve unit tests for OutputBuffer class

This commit modernizes the existing unit test for the OutputBuffer and
adds other three tests for the same class:
- outputBufferCheckCrcTest: to check CRC calculation from the bare
buffer.
- outputBufferCopyValueTest: to check the copy of a value from the
buffer and the CRC of it afterwards.
- outputBufferSeveralBlocksInBufferTest: to check the expected behavior
of the OutputBuffer when several blocks are added to it, with an
uncomplete last block. Header, CRC and data are checked in this test.

Signed-off-by: Dave <dave@leil.io>

lgsilva3087 commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -656,7 +659,7 @@ static void hddReadAheadAndBehind(IChunk *chunk, uint16_t block,
 		    blocksToBeReadAhead + block - firstBlockToRead);
 		OutputBuffer buffer =
-		    OutputBuffer(kHddBlockSize * (block - firstBlockToRead));
+		    OutputBuffer(1, block - firstBlockToRead);
```
Consider to use a constant like kIgnoreHeaderSize = 1.

dmga44 replied:
Done.

lgsilva3087 commented on src/chunkserver/buffers_pool.h:
```diff
@@ -44,46 +44,53 @@ class BuffersPool {
 
 private:
 	/// Maximum number of buffers in the pool.
 	static constexpr size_t kMaxSize = 8192;
```
Consider to use kMaxNumberOfBlocks instead of kMaxSize, for better control of the total memory size to be used by the buffer in the next steps (bigger size per write request).

Consider to log_trace the current size every defined amount of get or put operations (example every 1000 operations), to improve observability on memory usage.

dmga44 replied:
Done, both.

lgsilva3087 reviewed (commented):
Looks good, please see my small suggestions.

rolysr commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -744,24 +748,26 @@ int hddRead(uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
 
 		if (status == SAUNAFS_STATUS_OK) {
-			status = hddCheckCrcForFullBlock(chunk, block, outputBuffer, false);
+			status = hddCheckCrcForFullBlock(chunk, block, outputBuffer, 0, false);
```
You can consider in these cases, if the last params to be passed are 0 and false, then you can change the `hddCheckCrcForFullBlock` for having them as default on the signature definition.

dmga44 replied:
IMO I think it won't help too much, because this would be the only affeected case. Further changes would also add up calls to this function with starting offset different than 0, so I think it would be better to keep it explicit.

rolysr commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -744,24 +748,26 @@ int hddRead(uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
 		}
 	} else {  // Partial block
-		OutputBuffer tmp(kHddBlockSize);
+		OutputBuffer tmp(1, 1);
```
Consider to comment these magic `1` values here.

dmga44 replied:
The first parameter of the constructor was changed to `kIgnoreHeaderSize` to make it more readable, and added a comment on top to that line. The second `1` is because it is necessary to read a block.

rolysr commented on src/chunkserver/hddspacemgr.cc:
```diff
@@ -744,24 +748,26 @@ int hddRead(uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
 
 		if (status == SAUNAFS_STATUS_OK) {  // Successful read of the full block
-			status = hddCheckCrcForFullBlock(chunk, block, &tmp, true);
+			status = hddCheckCrcForFullBlock(chunk, block, &tmp, 0, true);
```
Same consideration here as in a previous comment.

dmga44 replied:
Note the first answer holds.

rolysr commented on src/chunkserver/output_buffer.cc:
```diff
@@ -30,71 +30,119 @@
-				return WRITE_AGAIN;
-			}
-			return WRITE_ERROR;
+			if (ret == 0 || errno == EAGAIN) { return WriteStatus::Again; }
```
Consider this change here:
```suggestion
			if (!ret || errno == EAGAIN) { return WriteStatus::Again; }
```

uristdwarf replied:
I'd rather be explicit and check for 0 explicitly, even though negatives convert to true. Seems more clear to me.

I don't know if it's logged in an above function, but even then I think you should log errno here (as an error), to better trace it later if needed.

uristdwarf replied:
Of course, if it's logged above, you probably need to remove or change the errno checking there, from man 3 errno:

```
      A common mistake is to do

           if (somecall() == -1) {
               printf("somecall() failed\n");
               if (errno == ...) { ... }
           }

       where errno no longer needs to have the value it had  upon  return  from
       somecall()  (i.e.,  it  may have been changed by the printf(3)).  If the
       value of errno should be preserved across a library  call,  it  must  be
       saved:

           if (somecall() == -1) {
               int errsv = errno;
               printf("somecall() failed\n");
               if (errsv == ...) { ... }
           }
```

dmga44 replied:
I agree with Urmas, I think it is better the explicit equal check with 0.

dmga44 replied:
Regarding @uristdwarf suggestion of logging errno here: the function which calls this `writeOutToAFileDescriptor` function logs an error using old LOG_NOTICE but does not specify. I could log it here, but then I'll have two logs about the same thing, and modifying it in an above function could already not have the initial errno.

So, I'll opt for leaving it like it is now.

rolysr commented on src/chunkserver/output_buffer.cc:
```diff
@@ -30,71 +30,119 @@
+	case BufferType::Header:
+		return headerBuffer_.copyIntoBuffer(mem, len);
+	default:
+		safs::log_err("Invalid buffer type");
```
Consider to improve the logging here by specifying that this comes from the OutputBuffer

dmga44 replied:
Done, I also changed the log type to warn because it is not critical, it is just ignored operation.

rolysr reviewed (approved):
LGTM.

uristdwarf commented on src/chunkserver/output_buffer.cc:
```diff
@@ -30,71 +30,119 @@
+	case BufferType::Header:
+		return headerBuffer_.copyIntoBuffer(mem.data(), mem.size());
+	default:
+		safs::log_err("Invalid buffer type");
+		return 0;
```
I wonder if an assert is more appropriate. In what situation would it be normal to have an invalid buffer type passed?

dmga44 replied:
I actually changed the log type to warn because it is not so critical, it is just an ignored operation.

Regarding your other question, it is almost impossible because those three types are the only defined ones. Currently all the calls to these functions only use those three buffers types.

uristdwarf reviewed (commented):
Review part 1/2?

dmga44 committed:
refactor: Address PR suggestions

Changes:
- keep track of the number of blocks in the BuffersPool instead of the
number of buffers.
- log the number of blocks in the BuffersPool every a fixed amount of
times with log_trace.
- define kIgnoreHeaderSize for the sizes of headers to be ignored.
- use the correct casting in writeOutToAFileDescriptor.
- use log_warn instead of log_err when wrong type of buffer if passed
in the OutputBuffer functions.

Signed-off-by: Dave <dave@leil.io>

Co-authored-by: guillex <guillex@leil.io>
Co-authored-by: rolysr <rolysr@leil.io>
Co-authored-by: Urmas Rist <urmas@leil.io>

lgsilva3087 reviewed (approved):
LGTM
